test/suites/clustering.sh

```diff
@@ -1052,3 +1052,113 @@ test_clustering_projects() {
   kill_lxd "${LXD_TWO_DIR}"
 }
+
+test_clustering_image_synchronization() {
+  setup_clustering_bridge
+  prefix="lxd$$"
+  bridge="${prefix}"
+
+  setup_clustering_netns 1
+  LXD_ONE_DIR=$(mktemp -d -p "${TEST_DIR}" XXX)
+  chmod +x "${LXD_ONE_DIR}"
+  ns1="${prefix}1"
+  spawn_lxd_and_bootstrap_cluster "${ns1}" "${bridge}" "${LXD_ONE_DIR}"
+
+  # Add a newline at the end of each line. YAML as weird rules..
+  cert=$(sed ':a;N;$!ba;s/\n/\n\n/g' "${LXD_ONE_DIR}/server.crt")
+
+  # Spawn a second node
+  setup_clustering_netns 2
+  LXD_TWO_DIR=$(mktemp -d -p "${TEST_DIR}" XXX)
+  chmod +x "${LXD_TWO_DIR}"
+  ns2="${prefix}2"
+  spawn_lxd_and_join_cluster "${ns2}" "${bridge}" "${cert}" 2 1 "${LXD_TWO_DIR}"
+
+  # Image synchronization is disabled by default
+  image_minimal_replica1=$(LXD_DIR="${LXD_ONE_DIR}" lxc config get cluster.image_minimal_replica)
+  image_minimal_replica2=$(LXD_DIR="${LXD_TWO_DIR}" lxc config get cluster.image_minimal_replica)
+  [ "$image_minimal_replica1" = "" ] || false
+  [ "$image_minimal_replica2" = "" ] || false
+
+  # Import the test image on node1
+  LXD_DIR="${LXD_ONE_DIR}" ensure_import_testimage
+
+  # The image is visible through both nodes
+  LXD_DIR="${LXD_ONE_DIR}" lxc image list | grep -q testimage
+  LXD_DIR="${LXD_TWO_DIR}" lxc image list | grep -q testimage
+
+  # The image tarball is only available on node1
+  fingerprint=$(LXD_DIR="${LXD_ONE_DIR}" lxc image info testimage | grep "Fingerprint:" | cut -f2 -d" ")
+  [ -f "${LXD_ONE_DIR}/images/${fingerprint}" ] || false
+  [ ! -f "${LXD_TWO_DIR}/images/${fingerprint}" ] || false
+
+  # Delete the imported image
+  LXD_DIR="${LXD_ONE_DIR}" lxc image delete testimage
+  [ ! -f "${LXD_ONE_DIR}/images/${fingerprint}" ] || false
+  [ ! -f "${LXD_TWO_DIR}/images/${fingerprint}" ] || false
+
+  # Enable the image synchronization across the nodes in the cluster.
+  LXD_DIR="${LXD_TWO_DIR}" lxc config set cluster.image_minimal_replica 2
+  LXD_DIR="${LXD_ONE_DIR}" lxc info | grep -q 'cluster.image_minimal_replica: "2"'
+  LXD_DIR="${LXD_TWO_DIR}" lxc info | grep -q 'cluster.image_minimal_replica: "2"'
+
+  # Import the test image on node2
+  LXD_DIR="${LXD_TWO_DIR}" ensure_import_testimage
+
+  # The image is visible through both nodes
+  LXD_DIR="${LXD_ONE_DIR}" lxc image list | grep -q testimage
+  LXD_DIR="${LXD_TWO_DIR}" lxc image list | grep -q testimage
+
+  # The image tarball is available on both node
+  fingerprint=$(LXD_DIR="${LXD_ONE_DIR}" lxc image info testimage | grep "Fingerprint:" | cut -f2 -d" ")
+  [ -f "${LXD_ONE_DIR}/images/${fingerprint}" ] || false
+  [ -f "${LXD_TWO_DIR}/images/${fingerprint}" ] || false
+
+  # Delete the imported image
+  LXD_DIR="${LXD_ONE_DIR}" lxc image delete testimage
+  [ ! -f "${LXD_ONE_DIR}/images/${fingerprint}" ] || false
+  [ ! -f "${LXD_TWO_DIR}/images/${fingerprint}" ] || false
+
+  # Import the image from a URL
+  LXD_DIR="${LXD_TWO_DIR}" lxc image import https://dl.stgraber.org/lxd --alias busybox-amd64
```

Comment:
I'm not sure we really want to test an actual URL, because it means the test needs to go on the network. At least, this should be executed conditionally only if the ```$LXD_OFFLINE``` environment variable is not ```1```.

@stgraber what do you think? should we keep this particular part of the test or just drop it?

I'd probably drop it.

Reply:
Yeah, I'd say let's drop that one. We don't have network access in most cases for our tests and when we do, there's no real guarantee that this particular URL will keep working.